### PR TITLE
fix(input): preserve text in DEL-coalesced chunks

### PR DESCRIPTION
**Pull request by chioarub**
## Summary

- process filtered terminal chunks containing raw DEL bytes and printable text in source order
- preserve grapheme-aware and token-aware deletion, cursor state, Vim replay, and coalesced Enter submission
- carry same-event bash mode context through PromptInput and REPL without changing ordinary Backspace/Delete handling

The SSH/tmux raw-DEL workaround previously counted DEL bytes, applied only those deletions, and returned. Terminal or IME replacement chunks such as `\x7fă` therefore discarded the printable replacement text. This change routes every printable segment through the existing insertion semantics and applies DEL runs between segments through the existing cursor abstraction.

## Impact

- user-facing impact: replacement text delivered alongside raw DEL bytes is no longer lost, including Unicode graphemes and interleaved chunks
- developer/maintainer impact: the transformation is isolated in deterministic helpers with coverage for filters, tokens, mode changes, Vim batching, and submission state

Most of the diff is regression coverage for the terminal input matrix and same-event React/Vim state transitions; no dependency, timeout, parser, or normalization policy changed.

## Testing

- [x] `bun run build`
- [x] `bun run smoke`
- [x] `bun run check`
- [x] focused tests: `bun test src/components/PromptInput/utils.test.ts src/components/TextInput.test.tsx src/hooks/useTextInput.test.ts src/hooks/usePasteHandler.test.ts src/utils/Cursor.nfc.test.ts src/components/PromptInput/inputModes.test.ts` — 80 passed
- [x] `bun run typecheck`
- [x] `bun run typecheck:type-tests`
- [x] `bun run test:provider` — 1,479 passed
- [x] `npm run test:provider-recommendation` — 129 passed
- [x] `bun run security:pr-scan -- --base upstream/main --head HEAD`
- [x] `bun run web:typecheck` and `bun run web:build`

The local full suite completed with 7,844 passing and 2 skipped tests. Its 41 failures are environment-dependent baseline failures reproduced on the untouched base; the exact upstream base SHA has green hosted smoke-and-tests, typecheck, and web jobs.

## Notes

- provider/model path tested: not applicable; the input transformation is provider-independent
- screenshots attached: not applicable; this corrects input behavior without visual presentation changes
- reviewer focus: `applyCoalescedDelInput`, bulk token-aware deletion equivalence, and same-event mode submission override
- known limitation: physical terminals may fragment or decorate IME byte streams differently; deterministic raw and ANSI-bearing chunks are covered without broadening global timeouts
- AI assistance was used during implementation and adversarial review; all accepted findings were verified and fixed before submission

## Summary by CodeRabbit

* **New Features**
  * Improved prompt submission across input modes, including mode-aware history and command handling.
  * Prompt suggestions and autocomplete now activate only in supported modes.
  * Enhanced text editing with reliable multi-character deletion and cursor preservation.

* **Bug Fixes**
  * Fixed combined Enter and Delete input, including SSH and Unicode scenarios.
  * Improved Vim insert-mode deletion, repeat behavior, and live-cursor updates.
  * Improved tab, spacing, help-toggle, and pasted-text handling.
  * Prevented unintended submissions from unsupported suggestion contexts.

### CODE REVIEW

**Comment by chioarub** [comment]
## Update

Addressed five substantiated review items in b49fc052 by expanding edge-case coverage, replacing fixed test sleeps with event predicates, reusing the production change-context type, and making kill/yank cleanup failure-safe. Production behavior is unchanged by this follow-up.

### Addressed

- **Prompt utility branch coverage** (review) — Added direct coverage for End-key classification and the no-pending-mode submission path. — b49fc052
- **Fixed post-input sleeps in TextInput harnesses** (review) — Replaced both 75 ms sleeps with a bounded predicate wait tied to receipt of the input event, including filter-rejected chunks. — b49fc052
- **Duplicated TextInput change-context type** (review) — Removed the local alias and imported TextInputChangeContext from the production type module. — b49fc052
- **Kill-ring and yank-state cleanup** (review) — Scoped the stateful tests under beforeEach/afterEach cleanup so a failed assertion cannot leak module-global state. — b49fc052
- **Rejected insertion followed by DEL** (review) — Added a direct helper regression proving a later DEL restores shouldCommit after an insertion callback returns undefined. — b49fc052

### Not changed

- **Additional extraction of the coalesced-DEL hook branch** (review) — The deterministic sequential transformation is already isolated in applyCoalescedDelInput and directly tested. The remaining branch coordinates hook-owned mode notifications, cursor mirroring, commits, and submission; extracting it in this review follow-up would broaden the diff without changing behavior. I can make that structural change if maintainers prefer it.

Validation:

- `bun test src/components/PromptInput/utils.test.ts src/components/PromptInput/inputModes.test.ts src/components/TextInput.test.tsx src/hooks/useTextInput.test.ts src/utils/Cursor.nfc.test.ts` — 75 pass, 0 fail
- `bun run typecheck` — passed
- `bun run typecheck:type-tests` — passed (10 files checked)
- `git diff --check` — passed

All five hosted checks were green and the PR was mergeable before this update. The refreshed upstream main remains an ancestor of the branch. This follow-up was prepared with AI assistance and manually verified against each live review thread.

**Comment by chioarub** [comment]
## Update

Addressed the new test-harness cleanup finding in 2eea887b. Both input scenario helpers now release their Ink root and streams even when a predicate wait times out; production behavior remains unchanged.

### Addressed

- **Harness cleanup after predicate timeouts** (review) — Wrapped the render, input, and wait sequence in each TextInput scenario helper with try/finally so root.unmount(), stdin.end(), and stdout.end() run on success or failure. — 2eea887b

Validation:

- `bun test src/components/TextInput.test.tsx` — 32 pass, 0 fail
- `bun run typecheck` — passed
- `bun run typecheck:type-tests` — passed (10 files checked)
- `git diff --check` — passed

All five hosted checks were green and the PR was mergeable before this push. Upstream main remains an ancestor of the branch. This follow-up was prepared with AI assistance and manually verified against the live review thread.

**Review by kevincodex1 — DISMISSED** [review]
looks great!

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P3] Restore trailing-CR stripping for embedded multiline paste
  `src/hooks/useTextInput.ts:54-57`
  Refactoring coalesced-Enter handling into `prepareTextInputEvent` dropped the old default-handler step that always stripped a final `\r` before converting remaining `\r` to `\n`. Inputs such as `a\rb\r` now become `a\nb\n` instead of `a\nb`. The PR claims no normalization/parser policy change, but this is a real (if narrow) regression for unbracketed multiline paste chunks that end with Enter. Please preserve the previous `(?<=[^\\\r\n])\r$` strip, or equivalent, before the global `\r`→`\n` conversion when `shouldSubmit` is false.

- [P3] Snapshot the pre-chunk buffer for undo on coalesced DEL + mode entry
  `src/components/PromptInput/PromptInput.tsx:889-903`
  When bash mode is entered inside a raw-DEL batch (`a` + `\x7f!` / `\x7f!ls\r`), `changeContext.previousValue` is the post-DEL segment cursor (`''`), and `pushToBuffer(previousValue, previousCursorOffset, …)` records that empty snapshot instead of the pre-event controlled input (`a`). Ctrl+Z / input-buffer undo cannot restore the text cleared by the coalesced DEL. This is new surface introduced by the fix, not a regression on `main`, but it is still a real gap in the new path. Please keep `changeContext` for `detectModeEntry`, but use the closure `input` / `cursorOffset` (or another pre-chunk snapshot) for `pushToBuffer`.

**Comment by chioarub** [comment]
## Update

Rebased onto the current `main` and fixed both requested input edge cases in `5a19b2e6`. The same push adds focused coverage for the related cursor, Vim repeat, tab, image-pill, and help-toggle paths.

### Addressed

- **Trailing CR in embedded multiline paste** (top-level-review) — A final carriage return is stripped after preserving embedded carriage returns as newlines, so `a\rb\r` now becomes `a\nb` without submitting. — 5a19b2e6
- **Mode-entry undo snapshot** (top-level-review) — Mode entry now snapshots the pre-chunk controlled input and cursor state instead of the segment-local change context. — 5a19b2e6
- **Rebase onto current main** (conflict) — Rebased the branch onto `5844d1fe`; the PromptInput and REPL changes remained conflict-free.

Validation passed: the focused six-file matrix (92 tests), `bun run typecheck`, `bun run typecheck:type-tests`, `bun run smoke`, `bun run security:pr-scan -- --base upstream/main --head HEAD`, and `git diff --check`. `bun run check` still exits on the repository's local order/environment-sensitive suite noise (7,859 pass, 2 skip, 41 fail); the exact upstream baseline produced 7,808 pass, 2 skip, 29 fail in the same environment.

File: src/components/TextInput.test.tsx [hunk 1]
```diff
@@ -377,3 +411,894 @@ test('VimTextInput preserves rapid typed characters before delayed parent value
...
+  inputState!.onInput('!\x7f\r', {} as Key)
+  await Bun.sleep(25)
+
+  root.unmount()
+  stdin.end()
+  stdout.end()
+
+  expect(observedMode).toBe('bash')
+  expect(submissions).toEqual([{ value: '', mode: 'bash' }])
+})
```
[the remark this answers is not in the document]
Comment (chioarub) [reply src/components/TextInput.test.tsx:994]: ## Update

Addressed the hook-test synchronization and cleanup concern.

### Addressed

- **Replace fixed sleeps and guarantee stream cleanup** — Replaced the six fixed 25 ms sleeps with predicate-based waits for the exact state transition under test, and moved each root/stream cleanup sequence into a finally block so assertion failures cannot leak resources. — 88764814

The updated TextInput suite passes all 36 tests, including repeated focused execution before this update.

File: src/hooks/useTextInput.ts [hunk 2]
```diff
@@ -36,9 +37,125 @@ function mapInput(input_map: Array<[string, InputHandler]>): InputMapper {
...
+    cursor.text.length === 0 &&
+    cursor.isAtStart() &&
+    isInputModeCharacter(input)
+  ) {
+    options.onModeCharacter?.(text, {
+      previousValue: cursor.text,
+      cursorOffset: cursor.offset,
+    })
+    return undefined
+  }
```
[the remark this answers is not in the document]
Comment (chioarub) [reply src/hooks/useTextInput.ts:99]: ## Update

Addressed ANSI-wrapped mode-character detection.

### Addressed

- **Classify the stripped printable input** — applyPrintableInput now checks the VT-stripped text when deciding whether the input is a mode character. A focused regression verifies that an ANSI-wrapped exclamation mark enters the mode path instead of being inserted literally. — 88764814

The new test fails on the prior PR head and passes with this commit.

File: src/hooks/useTextInput.ts [hunk 3]
```diff
@@ -540,25 +631,108 @@ export function useTextInput({
...
+          if (previousOffset !== result.cursor.offset) {
+            onOffsetChange(result.cursor.offset)
+          }
+        }
       }
       resetKillAccumulation()
       resetYankState()
+      if (result.shouldCommit && preparedInput.shouldSubmit) {
+        onSubmit?.(result.cursor.text)
+      }
```
[the remark this answers is not in the document]
Comment (chioarub) [reply src/hooks/useTextInput.ts:757]: ## Update

Kept submission on the final transformed cursor deliberately.

### Not changed

- **Do not submit the synthetic mode sentinel** — committedCursor can contain a synthetic mode sentinel whose only purpose is to notify onChange of mode entry. Submitting that cursor would resurrect a character already consumed by PromptInput. The raw submission regressions now assert that consuming the sentinel submits `a`, while typing a later literal sentinel submits `!a`; both would regress if committedCursor were submitted. — 88764814

A source comment now documents this boundary, and same-batch editing is synchronized independently without changing the submission contract.

**Comment by chioarub** [comment]
## Update

Verified the outside-diff Vim ANSI note against the current head.

### Not changed

- **Additional ANSI stripping in the DEL recording branch** — The non-mode segment is already passed through applyPrintableInput when it is added to insertedTextCursor, and applyPrintableInput strips VT control characters before insertion. I reproduced the exact DEL plus ANSI-wrapped `b`, Escape, dot-repeat path: it produces `bb` and contains neither DEL nor ESC bytes. An additional strip at this call site would therefore be redundant. — 88764814

No code change was made for this non-reproducing advisory note.

**Review by jatmn — APPROVED** [review]
@kevincodex1 LGTM

**Review by kevincodex1 — APPROVED** [review]
lgtm
